net_util/src/tap.rs

```diff
@@ -400,25 +397,36 @@ mod tests {
     // Given a buffer of appropriate size, this fills in the relevant fields based on the
     // provided information. Payload refers to the UDP payload.
     fn pnet_build_packet(buf: &mut [u8], dst_mac: MacAddr, payload: &[u8]) {
-        let mut eth = MutableEthernetPacket::new(buf).unwrap();
-        eth.set_source(MacAddr::new(0x06, 0, 0, 0, 0, 0));
-        eth.set_destination(dst_mac);
-        eth.set_ethertype(EtherTypes::Ipv4);
-
-        let mut ipv4 = MutableIpv4Packet::new(eth.payload_mut()).unwrap();
-        ipv4.set_version(4);
-        ipv4.set_header_length(5);
-        ipv4.set_total_length(20 + 8 + payload.len() as u16);
-        ipv4.set_ttl(200);
-        ipv4.set_next_level_protocol(IpNextHeaderProtocols::Udp);
-        ipv4.set_source(Ipv4Addr::new(192, 168, 241, 1));
-        ipv4.set_destination(Ipv4Addr::new(192, 168, 241, 2));
-
-        let mut udp = MutableUdpPacket::new(ipv4.payload_mut()).unwrap();
-        udp.set_source(1000);
-        udp.set_destination(1001);
-        udp.set_length(8 + payload.len() as u16);
-        udp.set_payload(payload);
+        // Make a new vector that can accommodate the header (buf) and data (payload)
+        // we can't escape the copy here, because from_bytes consumes buf
+        // unless we change the ether API in dumbo
+        let mut v = vec![0u8; buf.len()];
+        let mut eth = EthernetFrame::from_bytes(v.as_mut_slice()).unwrap();
+
+        let src_mac = MacAddr::from_bytes(&[0x06, 0, 0, 0, 0, 0]).unwrap();
+        eth.set_src_mac(src_mac);
+        eth.set_dst_mac(dst_mac);
+        eth.set_ethertype(ETHERTYPE_IPV4);
+
+        // Because we're borrowing eth as mutable
+        {
+            let mut ipv4 = IPv4Packet::from_bytes_unchecked(eth.payload_mut());
+            let ip_header_len_bytes = 20;
+            ipv4.set_version_and_header_len(IPV4_VERSION, ip_header_len_bytes);
```

Comment:
These function can be chained; i.e. can call `ipv4.set_version_and_header_len(..).set_total_len(..)...` etc, because they return a `&mut Self`.